ENH: find and replace functionality

tangkong opened the pull request:
## Description
- Implements atef-specific find and replace functionality
- adds a basic find-and-replace widget
- finish up / clean up widget button functionality

## Motivation and Context
This stems from a request for templated checkouts, which boils down to having find-and-replace functionality.  A naive find-and-replace on the json could touch file/field names, and is rather difficult to inspect.  

My intention is to make the future "fill template" step / tool wrap this functionality, possibly with some extra visualization or streamlining buttons.  A sample checkout could be run through this tool and and "filled" as is with this tool, I imagine

(In reality I implemented the simple find-and-replace helpers, tried making it more robust, and a week later ended up here.)

Creating a step that fills templates mid-checkout is a possible next step, though we'd have to consider expanding each template step on preparation and inserting it into the tree.  Inspection of the substeps of that filled template would be un-inspectable unless we expand in edit-mode, and in that case we're basically creating a bunch of filled template checkouts separately.

## How Has This Been Tested?
Some unit tests have been added, more need to be...

## Where Has This Been Documented?
This PR

## Screenshots 
![image](https://github.com/pcdshub/atef/assets/35379409/b88ccb99-49c4-4297-ae9b-0972d17218d2)

![image](https://github.com/pcdshub/atef/assets/35379409/d4259d8e-fe36-4868-a81b-555dab66dc29)

![image](https://github.com/pcdshub/atef/assets/35379409/1ec61c27-81fd-4946-970a-f1462910d16d)


## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive docstrings, including context and API
- [x] New/changed functions and methods are covered in the test suite where possible
- [x] Code has been checked for threading issues (no blocking tasks in GUI thread)
- [x] Test suite passes locally
- [x] Test suite passes on GitHub Actions
- [x] Ran ``docs/pre-release-notes.sh`` and created a pre-release documentation page
- [x] Pre-release docs include context, functional descriptions, and contributors as appropriate

tangkong commented:
This is basically feature complete as far as I care for.  There was a good amount of gui-ing here so I'm hoping for some adversarial clicking while I write comments / unit tests.

ZLLentz commented on atef/ui/fill_template_page.ui:
Clickly clicky seemed to work as I used this and it seems very powerful in a way that I think will be well-appreciated by the users. A few thoughts:
- Subbing into a big structure can sometimes show a bit too much info. Is there any way we can truncate this so we only show the bit that changed?
- Is there any simple way we can automatically refresh when I add a new edit? I got confused for a moment at first before realizing that I needed to click the refresh button on the left.

![image](https://github.com/pcdshub/atef/assets/10647860/12e2c777-895d-4be5-b1c8-b896489a846f)

ZLLentz replied:
Oh, also: after saving the file, it seems like this whole template window goes blank. I think that will scare some people. I wonder if we should automatically close this and open the file that was just saved?

tangkong replied:
> Subbing into a big structure can sometimes show a bit too much info. Is there any way we can truncate this so we only show the bit that changed?

I gave this my best effort, but I wonder if I can do any better.  Currently it shows the deepest dataclass, which I thought was needed to give the change any context at all.  But I suppose now that I have {object}.{attr} on the left I could just show the {attr} on the right

> Is there any simple way we can automatically refresh when I add a new edit? I got confused for a moment at first before realizing that I needed to click the refresh button on the left.

This is a perpetual issue I have.  I always try to hook more `editingFinished` signals, but find them to either not catch all the cases I'd want or trigger too often.  I'll see what I can do here

tangkong replied:
I've made the previews show just the bit that changed (the last part of the path), as well as gave the user some notice that their file was saved and that they can open it.  

I'll give a bit more information in the details list if the user hasn't picked an edit as well

tangkong replied:
I did think about this, but punted it never to look back.  I think it's honestly the ideal visualization, but involves quite a bit of work I think.  (Particularly because the tree does not often match the dataclass structure)

This is definitely worth a follow up issue methinks

ZLLentz reviewed (approved):
I'm very positive on this. Pending Ken's review I think this is good enough to merge.

ZLLentz commented on atef/widgets/config/window.py:
```diff
@@ -367,13 +374,39 @@ def clear_results(self, *args, **kwargs):
             current_tree.update_statuses()
 
+    def find_replace(self, *args, **kwargs):
+        """ find and replace in the current tree.  Open a find-replace widget """
```
this is really nifty

I wonder if anyone will get confused by the interface though- it wasn't obvious to me at first that this wasn't doing a find/replace in the same window, rather it is creating a whole new version that I could choose to edit and save seperately

klauer replied:
Yeah, it is really cool but I also felt the same regarding the end user's expectations with the dialog.

"Apply All" seems like it should be replacing it in the file, but it's just doing the changes on an in-memory version until you click 'save as' (... right?)
Maybe the dialog should be something like:
> "Choose input file: "
> "Choose output file: "
> (Make edits in this section)
> Reload / Preview and Verify / Save

tangkong commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,992 @@
+        preview_file = copy.deepcopy(self.data.target)
+
+        try:
+            _ = data.apply(target=preview_file)
```
Ah I dropped my train of thought here.  `FindReplaceAction.apply` catches errors now, so I need that value to decide whether or not application was successful

tangkong commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,992 @@
+
+            self.actions.append(action)
+
+        # this works, but is smelly and bad.  Access parent table signals
```
The caveat here is that the `TableWidgetWithAddRow` takes care of adding rows to the point where it never exposes those RowWidgets.

Adding a signal on this row that the table widget expects would force me to change the other instances of this table-row-thing to match, which I felt wasn't worth it at this juncture.  I totally agree though, this isn't quite how things should be done.

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+            return
+        elif not edit_row_widget.get_details_rows():
+            l_item = QtWidgets.QListWidgetItem(
+                'provide search and replace text to show details...'
```
Hmm, it wasn't clear to me why the "edit details" weren't showing:
<img width="803" alt="image" src="https://github.com/pcdshub/atef/assets/5139267/beb52117-d9d6-439f-aed1-a1e21dc803c1">
I think this should probably read "Click on edit in the left list to see details; if nothing shows here there is no match" or something to that effect

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+        refresh_button.setToolTip('refresh edit details')
+        refresh_button.setIcon(qta.icon('ei.refresh'))
+
+        apply_button = self.button_box.button(QtWidgets.QDialogButtonBox.Apply)
```
If I click 'apply all' after having changed a device in the template, I'd have expected to see the device/signal list above change:
<img width="532" alt="image" src="https://github.com/pcdshub/atef/assets/5139267/6b5a76c6-8e5a-4a54-9e13-9211f638994a">

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+    file_name_label: QtWidgets.QLabel
+    type_label: QtWidgets.QLabel
+
+    device_table: QtWidgets.QTableWidget
```
I think the biggest feature you could add is the ability to change (add/remove/edit) devices and signals directly in this table using the device/component selector dialog:
<img width="343" alt="image" src="https://github.com/pcdshub/atef/assets/5139267/53c01d81-1f49-402d-8e21-6faef0c5bd62">

The find/replace stuff is very valuable, but I think the primary thing people will want to template is "test that device instead of this device"

tangkong replied:
aaaah now that's a fancy idea.  I'll make a followup, because my vision of this quickly became very complicated

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+        new_enum = getattr(final_step[1], replace_fn(old_enum.name))
+        setattr(parent_item, path[-2][1], new_enum)
+    else:
+        # simple field paths don't have a final (__sth__, ?) segement
```
```suggestion
        # simple field paths don't have a final (__sth__, ?) segment
```

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+        for d_key, d_value in item.items():
+            # don't halt at first key match, values could also have matches
+            if isinstance(d_value, get_args(PrimitiveType)) and match(d_value):
+                yield parent + [('__dictvalue__', d_key)]
```
Checking the coverage - running test suite locally, it doesn't touch this line
(`coverage html && open htmlcov/index.html`)

GUI stuff aside, in my opinion we should try to aim for ~100% coverage on core library stuff that deals with our dataclasses. (Easier said than done of course!)

tangkong replied:
I should do this more regularly in my code review as well as devlopment

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+
+
+def verify_changes(parent, orig_file: Union[ConfigurationFile, ProcedureFile]) -> None:
+    # check to make sure changes are valid
```
```suggestion
    """Check to make sure changes are valid."""
```

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+        self,
+        target: Optional[Union[ConfigurationFile, ProcedureFile]] = None,
+        path: Optional[List[Tuple[Any, Any]]] = None,
+        replace_fn: Optional[Callable] = None
```
```python
ReplaceFunction = Callable[...something here...]
```
and
```suggestion
        replace_fn: Optional[ReplaceFunction] = None
```

Would recommend typing the "replace_fn" so you can tell if it's valid

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+        Parameters
+        ----------
+        target : Optional[Union[ConfigurationFile, ProcedureFile]], optional
+            the file to apply the find-replace action to, by default None
```
```suggestion
            The file to apply the find-replace action to, by default this applies to the current target of the action
```

Or something like that?

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+
+        return data
+
+    def update_replace_fn(self, *args, **kwargs) -> None:
```
Abbreviated method name with star args/kwargs and no docstring... Sorry to be picky (it's genetic, so you're supposed to excuse it I think), but would you mind either expanding out the name or adding brief intent docstrings to these?

tangkong replied:
Don't be sorry, it's entirely your prerogative to to complain about this.  I should be less lazy.

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+    def verify_changes(self) -> None:
+        verify_changes(self, self.orig_file)
+
+    def save_file(self) -> None:
```
Shouldn't there be an implicit "verify_changes" call before it's saved?

tangkong replied:
I agree with verifying as long as it doesn't prevent saving, since users might use this in an environment where they don't have access to all the necessary devices, making preparation fail.

klauer replied:
Something like "Verification failed! Save anyway? [y/n]"

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+            self.device_table.setItem(i, 0, QtWidgets.QTableWidgetItem(dev))
+
+    def verify_changes(self) -> None:
+        verify_changes(self, self.orig_file)
```
Careful, `self.orig_file` can be `None` here per the static analyzer

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+                data = deserialize(ProcedureFile, self._original_json)
+            except ValidationError:
+                logger.error('failed to open file as either active '
+                             'or passive checkout')
```
This should probably `raise` and clear `original_json` 
At the very least `data` is unbound in this except clause

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+        self,
+        *args,
+        filepath: Optional[str] = None,
+        window: Optional[Any] = None,
```
Probably can do better than `Any` here

tangkong replied:
TIL `typing.TYPE_CHECKING` was a thing.  circular imports be gone!

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+    def __init__(
+        self,
+        *args,
+        data=None,
```
`data` unused?

tangkong replied:
My `TableWidgetWithAddRow` expected its row widgets to be `DataWidget`s, all of which have a `data` kwarg.  I probably could have made a dataclass to represent an "Edit" but apparently past me didn't feel like it 🤷

klauer commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+        """
+        Update the title.  Will be the name and the number of unapplied edits
+        """
+        file_name = os.path.basename(self.fp)
```
Dealing with `None` is annoying with respect to the static analyzer but I find it to be good practice - even if you're 100% sure this doesn't get called when `fp is None` - to guard for it anyway. A future refactor you may forget about it and then get confused why `basename(None)` is getting called

klauer commented on atef/widgets/utils.py:
```diff
@@ -216,3 +216,93 @@ def reset_cursor(self):
+        self.toggle_button = QtWidgets.QToolButton(text=text)
+        self.toggle_button.setCheckable(True)
+        self.toggle_button.setChecked(False)
+        self.toggle_button.setStyleSheet("QToolButton { border: none; }")
```
Aside: we should probably move to making an atef stylesheet at some point and move these things there (`atef.qss`, oh yeah!)

tangkong replied:
I wonder if that'd fix some of the weird osx rendering differences 🤔 A good idea, but I'm not volunteering right now... 👀

klauer reviewed (approved):
The functionality here is awesome! Nicely done. I think this will serve the end user rather well.

Some thoughts, suggestions, nitpicks, and musings as per usual. Nothing at a level that requires holding this PR up any longer. Sorry it took me so long to get around to reviewing!

tangkong commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+                yield parent + [('__list__', idx)]
+            else:
+                yield from walk_find_match(l_item, match,
+                                           parent=parent + [('__list__', idx)])
```
This would have probably made things easier.  I'll choose to pass on this, but toss it into an issue if later me decides I have extra cycles

tangkong commented on atef/widgets/config/find_replace.py:
```diff
@@ -0,0 +1,1009 @@
+
+    def load_file(self, filepath) -> Union[ConfigurationFile, ProcedureFile]:
+        with open(filepath, 'r') as fp:
+            self._original_json = json.load(fp)
```
Good catch, this is left over from when I was editing the json directly.  Both here and in the FindReplaceWidget, the original json is never used.

tangkong replied:
Also you're totally right we do have loaders. 🤦

tangkong commented:
I think I've addressed all the comments that fit as add-ons to this PR, and made issues for the rest.  I'd like to merge this and move on, unless there are any objections

klauer commented:
Merge at will!
